File: Microservices/clients_service/src/main/java/clients/clients_service/service/ClientService.java
package clients.clients_service.service;

import clients.clients_service.entity.Client;

import java.util.List;
import java.util.Optional;

public interface ClientService {
    Client createClient(Client client);
    List<Client> getAllClients();
    Optional<Client> getClientById(Long id);
    Client updateClient(Long id, Client clientDetails);
    void deleteClient(Long id);
    Optional<Client> getClientByCne(String cne);
    public void sendClient(Client client);
   public Client getClientByIdd(Long idClient);
    public void sendClientfordeletion(Client client);
    public void sendClientforupdate(Client client);


}

File: frontend/front/src/app/services/client.service.ts
// src/app/services/client.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Client } from '../models/client.model';

@Injectable({
  providedIn: 'root',
})
export class ClientService {
  private apiUrl = 'http://localhost:8080/api/clients';

  constructor(private http: HttpClient) {}

  getClients(): Observable<Client[]> {
    return this.http.get<Client[]>(this.apiUrl);
  }

  getClientById(id: number): Observable<Client> {
    return this.http.get<Client>(`${this.apiUrl}/${id}`);
  }

  createClient(client: Client, loggedInUser: string): Observable<Client> {
    const headers = new HttpHeaders().set('loggedInUser', loggedInUser);
    return this.http.post<Client>(this.apiUrl, client, { headers });
  }

  updateClient(id: number, client: Client): Observable<Client> {
    return this.http.put<Client>(`${this.apiUrl}/${id}`, client);
  }

  deleteClient(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getClientByCne(cne: string): Observable<Client> {
    return this.http.get<Client>(`${this.apiUrl}/cne/${cne}`);
  }
}
